mobile/src/employee-details/event-action-provider.ts

```diff
@@ -0,0 +1,120 @@
+    readonly name: string;
+}
+
+//============================================================================
+export interface EventActionsContainer {
+    readonly event: CalendarEvent;
+    readonly employee: Employee;
+    readonly positiveAction?: EventAction;
+    readonly negativeAction?: EventAction;
+}
+
+//============================================================================
+export class EventActionProvider {
+    //----------------------------------------------------------------------------
+    constructor(private readonly userId: EmployeeId,
+                private readonly eventSetStatus: (employeeId: EmployeeId, calendarEvent: CalendarEvent, status: CalendarEventStatus) => void,
+                private readonly eventApprove: (employeeId: EmployeeId, calendarEvent: CalendarEvent) => void) {
+    }
+
+    //----------------------------------------------------------------------------
+    public getEventActions(events: CalendarEvent[],
+                           employee: Employee,
+                           permissions: UserEmployeePermissions,
+                           approvals: Map<CalendarEventId, Set<Approval>>): EventActionsContainer[] {
+
+        const eventContainers: EventActionsContainer[] = [];
+
+        for (const event of events) {
+            eventContainers.push(this.evaluateEvent(event, employee, permissions, approvals));
+        }
+
+        return eventContainers;
+    }
+
+    //----------------------------------------------------------------------------
+    public getRequestActions(events: CalendarEvent[],
+                             employee: Employee,
+                             approvals: Map<CalendarEventId, Set<Approval>>): EventActionsContainer[] {
+
+        const permissions = new UserEmployeePermissions();
+        permissions.permissionsNames = Set([Permission.approveCalendarEvents, Permission.rejectCalendarEvents]);
+        return this.getEventActions(events, employee, permissions, approvals);
+    }
+
+    //----------------------------------------------------------------------------
+    private evaluateEvent(event: CalendarEvent,
+                          employee: Employee,
+                          permissions: UserEmployeePermissions,
+                          approvals: Map<CalendarEventId, Set<Approval>>): EventActionsContainer {
+        let positiveAction: Optional<EventAction>;
+        let negativeAction: Optional<EventAction>;
+
+        const isOwn = this.userId === employee.employeeId;
+
+        if (event.status === CalendarEventStatus.Requested) {
+            if (permissions.has(Permission.approveCalendarEvents)) {
+                if (!this.approvedByMe(event, approvals)) {
+                    positiveAction = {
+                        handler: () => this.eventApprove(employee.employeeId, event),
+                        name: 'approve',
+                    };
+                }
+            }
+
+            if (isOwn) {
```

Comment:
This condition looks monstrous. Consider refactoring.

Reply:
Brobdingnagian? -)))

Reply:
Done. da1915058ebc114dd2c5cc343515801e95e2c1fb